Add Embroider support

simonihmig opened the pull request:
Given that `postprocessTree('all', tree)` is not supported in an Embroider-based build, this adds a `prerender` export that users can use to wrap their final app tree to add the prerendered HTML files, in a way as suggested in https://github.com/ef4/prember/issues/59#issuecomment-720225598:

```
// ember-cli-build.js
const { Webpack } = require('@embroider/webpack');
const compiledApp = require('@embroider/compat').compatBuild(app, Webpack, {
  // ...
});

return require('prember').prerender(app, compiledApp);
```

~This is still a WIP, would like to get feedback/approval for that approach first, then:~
* ~add some documentation~
* ~setup Embroider test scenarios, after #63 is merged~
* ~rebase when #61 is merged (this is still included here)~

Closes #59

simonihmig commented:
Rebased on top of merged #63 and added embroider scenarios to CI. Which pass! :tada:

josemarluedke commented:
@simonihmig Now that #61 is merged, do you want to rebase this PR and see if we can land it? Looking forward to using prember + embroider. :D

simonihmig commented:
> Now that #61 is merged, do you want to rebase this PR and see if we can land it? Looking forward to using prember + embroider. :D

@josemarleudke thanks for pinging! Just rebased and updated `@embroider/test-setup` to 0.41.0. Cannot go all the way to the latest version, as that dropped support for node 10, which I don't want to do in this PR.

CI looks "good", Embroider tests have passed if you look into the logs of the CI jobs. Canary is failing, as it is currently for *all addons*, due to https://github.com/emberjs/ember-test-helpers/pull/1066 waiting to be released... 😕

So I moved this out of draft status for @ef4 to review! 🙂

villander commented on index.js:
```diff
@@ -7,8 +7,31 @@ module.exports = {
+   * don't support the `postprocessTree('all', tree)` hook used here.
+   */
+  prerender(app, tree) {
+    let premberAddon = app.project.addons.find(({ name }) => name === 'prember');
```
@simonihmig how about add test coverage for this one and `_prerenderTree`?

simonihmig replied:
The embroider scenarios are using the `prerender` to run the whole test suite, see https://github.com/ef4/prember/pull/64/files#diff-e835654135548f75db36136807171b4d12d3514e043f4f80a146b88cc668fba4R23. 

And `_prerenderTree` is a private method, and used both by `postprocessTree` and `prerender` (basically extracted from `postprocessTree` so it can be shared with `prerender`), so already implicitly covered by prember`s node-based acceptance tests. Or what kind of test did you have in mind?

simonihmig commented:
Canary is still failing, now with `ember-cli-fastboot` in its 2.2.3 release using the same [invalid import](https://github.com/ember-fastboot/ember-cli-fastboot/blob/114a2ad886c9f1cca4427c5dc67435903c1075be/addon/services/fastboot.js#L2). 

Same happens in prember's master branch (see [CI](https://github.com/ef4/prember/runs/3159416604?check_suite_focus=true#step:5:623)), so unrelated to this PR. 

This seems fixed in master of `ember-cli-fastboot`, but that has only 3.0 and 3.2 beta releases. Can/should we update? Haven't followed the developments in `ember-cli-fastboot`, not sure why there are only beta releases!?

ef4 commented:
Hi @simonihmig, I finally got #70 merged so this can get rebased and then I think it's good to go.

simonihmig commented:
@ef4 done!

simonihmig commented:
> add some documentation

This is still missing though... I was waiting for feedback whether the changes here are actually going to land as they are, before working on docs. Will add something shortly...

simonihmig commented:
> add some documentation

Did that now, hope that works!

ef4 commented:
Published 1.1.0.

sandstrom commented:
Awesome @simonihmig and @ef4 🎉
